SQL Assignment

Consider a meal delivery company which operates in multiple cities. They have various
fulfillment centers in these cities for dispatching meal orders to their customers.
They have the following data with them from all the fulfillment centers:

● Historical data of demand for a product-center combination (Weeks: 1 to 145)

● Product(Meal) features such as category, sub-category, current price and
discount

● Information for fulfillment center like center area, city information etc.


Assignment Questions
1. What are the distinct number of meal categories and cuisines?
2. Which center_id has the highest num_orders?
3. What is the top selling cuisine at the center_id that had the highest num_orders?
4. What is the average op_area per center_type?
5. Which center_type had the highest revenue? (Revenue is total sum of
checkout_price*num_orders)
6. Which is the top ordered cuisine in terms of num_orders?
7. What are the num_orders per cuisine per week?
8. Which center_id gave the highest number of discounts? (Discount is considered
when checkout_price is less than base_price)

In [0]:
%python
# File location and type
file_location_center = "/FileStore/tables/fulfilment_center_info.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df_center = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location_center)

In [0]:
%python 

temp_table_name = "fulfilment_center_info_csv"

df_center.createOrReplaceTempView(temp_table_name)

In [0]:
%sql

select * from `fulfilment_center_info_csv`

center_id city_code region_code center_type op_area 11 679 56 TYPE_A 3.7 13 590 56 TYPE_B 6.7 124 590 56 TYPE_C 4.0 66 648 34 TYPE_A 4.1 94 632 34 TYPE_C 3.6 64 553 77 TYPE_A 4.4 129 593 77 TYPE_A 3.9 139 693 34 TYPE_C 2.8 88 526 34 TYPE_A 4.1 143 562 77 TYPE_B 3.8 101 699 85 TYPE_C 2.8 86 699 85 TYPE_C 4.0 32 526 34 TYPE_A 3.8 149 478 77 TYPE_A 2.4 152 576 34 TYPE_B 4.0 92 526 34 TYPE_C 2.9 27 713 85 TYPE_A 4.5 14 654 56 TYPE_C 2.7 26 515 77 TYPE_C 3.0 104 647 56 TYPE_A 4.5 77 676 34 TYPE_A 3.8 23 698 23 TYPE_A 3.4 97 628 77 TYPE_A 4.6 146 526 34 TYPE_B 5.0 113 680 77 TYPE_C 4.0 145 620 77 TYPE_A 3.9 80 604 56 TYPE_C 5.1 55 647 56 TYPE_C 2.0 186 649 34 TYPE_A 3.4 99 596 71 TYPE_A 4.5 91 590 56 TYPE_C 0.9 20 522 56 TYPE_A 4.0 106 675 34 TYPE_A 4.0 81 526 34 TYPE_A 4.0 73 576 34 TYPE_A 4.0 29 526 34 TYPE_C 4.0 43 590 56 TYPE_A 5.1 102 593 77 TYPE_A 2.8 61 473 77 TYPE_A 4.5 50 556 77 TYPE_A 4.8 83 659 77 TYPE_A 5.3 57 541 77 TYPE_C 2.8 126 577 56 TYPE_A 2.7 177 683 56 TYPE_A 3.4 67 638 56 TYPE_B 7.0 174 700 56 TYPE_A 7.0 59 456 56 TYPE_A 4.2 58 695 77 TYPE_C 3.8 65 602 34 TYPE_A 4.8 39 526 34 TYPE_C 3.8 132 522 56 TYPE_A 3.9 89 703 56 TYPE_A 4.8 162 526 34 TYPE_C 2.0 75 651 77 TYPE_B 4.7 72 638 56 TYPE_C 3.9 41 590 56 TYPE_C 1.9 10 590 56 TYPE_B 6.3 110 485 77 TYPE_A 3.8 52 685 56 TYPE_B 5.6 93 461 34 TYPE_A 3.9 74 702 35 TYPE_A 2.8 34 615 34 TYPE_B 4.2 137 590 56 TYPE_A 4.4 153 590 56 TYPE_A 3.9 24 614 85 TYPE_B 3.6 109 599 56 TYPE_A 3.6 108 579 56 TYPE_B 4.4 36 517 56 TYPE_B 4.4 157 609 93 TYPE_A 4.1 17 517 56 TYPE_A 3.2 161 658 34 TYPE_B 3.9 42 561 77 TYPE_B 3.9 53 590 56 TYPE_A 3.8 30 604 56 TYPE_A 3.5 76 614 85 TYPE_A 3.0 68 676 34 TYPE_B 4.1 51 638 56 TYPE_A 7.0

In [0]:
%python
# File location and type
file_location_meal_info = "/FileStore/tables/meal_info.csv" 
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df_meal_info = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location_meal_info)


In [0]:
%python
# Create a view or table

temp_table_name = "meal_info_csv"

df_meal_info.createOrReplaceTempView(temp_table_name)

In [0]:
%sql

/* Query the created temp table in a SQL cell */

select * from `meal_info_csv`

meal_id category cuisine 1885 Beverages Thai 1993 Beverages Thai 2539 Beverages Thai 1248 Beverages Indian 2631 Beverages Indian 1311 Extras Thai 1062 Beverages Italian 1778 Beverages Italian 1803 Extras Thai 1198 Extras Thai 2707 Beverages Italian 1847 Soup Thai 1438 Soup Thai 2494 Soup Thai 2760 Other Snacks Thai 2490 Salad Italian 1109 Rice Bowl Indian 2290 Rice Bowl Indian 1525 Other Snacks Thai 2704 Other Snacks Thai 1878 Starters Thai 2640 Starters Thai 2577 Starters Thai 1754 Sandwich Italian 1971 Sandwich Italian 2306 Pasta Italian 2139 Beverages Indian 2826 Sandwich Italian 2664 Salad Italian 2569 Salad Italian 1230 Beverages Continental 1207 Beverages Continental 2322 Beverages Continental 2492 Desert Indian 1216 Pasta Italian 1727 Rice Bowl Indian 1902 Biryani Indian 1247 Biryani Indian 2304 Desert Indian 1543 Desert Indian 1770 Biryani Indian 2126 Pasta Italian 1558 Pizza Continental 2581 Pizza Continental 1962 Pizza Continental 1571 Fish Continental 2956 Fish Continental 2104 Fish Continental 2444 Seafood Continental 2867 Seafood Continental 1445 Seafood Continental

In [0]:
%python
# File location and type
file_location_train = "/FileStore/tables/train.csv" 
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df_train = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location_train)


In [0]:
%python
# Create a view or table

temp_table_name = "train_csv"

df_train.createOrReplaceTempView(temp_table_name)

In [0]:
%python

spark.sql(""" select * from `train_csv` """).show()

+-------+----+---------+-------+--------------+----------+---------------------+-----------------+----------+
 id|week|center_id|meal_id|checkout_price|base_price|emailer_for_promotion|homepage_featured|num_orders|
+-------+----+---------+-------+--------------+----------+---------------------+-----------------+----------+
1379560| 1| 55| 1885| 136.83| 152.29| 0| 0| 177|
1466964| 1| 55| 1993| 136.83| 135.83| 0| 0| 270|
1346989| 1| 55| 2539| 134.86| 135.86| 0| 0| 189|
1338232| 1| 55| 2139| 339.5| 437.53| 0| 0| 54|
1448490| 1| 55| 2631| 243.5| 242.5| 0| 0| 40|
1270037| 1| 55| 1248| 251.23| 252.23| 0| 0| 28|
1191377| 1| 55| 1778| 183.36| 184.36| 0| 0| 190|
1499955| 1| 55| 1062| 182.36| 183.36| 0| 0| 391|
1025244| 1| 55| 2707| 193.06| 192.06| 0| 0| 472|
1054194| 1| 55| 1207| 325.92| 384.18| 0| 1| 676|
1469367| 1| 55| 1230| 323.01| 390.0| 0| 1| 823|
1029333| 1| 55| 2322| 322.07| 388.0| 0| 1| 972|
1446016| 1| 55| 2290| 311.43| 310.43| 0| 0| 162|
1244647| 1| 55| 1727| 445.23| 446.23| 0| 0| 420|
1378227| 1| 55| 1109| 264.84| 297.79| 1| 0| 756|
1181556| 1| 55| 2640| 282.33| 281.33| 0| 0| 108|
1313873| 1| 55| 2306| 243.5| 340.53| 0| 0| 28|
1067069| 1| 55| 2126| 486.0| 485.0| 0| 0| 28|
1058482| 1| 55| 2826| 306.58| 305.58| 0| 0| 188|
1240935| 1| 55| 1754| 289.12| 289.12| 0| 0| 485|
+-------+----+---------+-------+--------------+----------+---------------------+-----------------+----------+
only showing top 20 rows

Joining all 3 tables into 1 using Nested Join

In [0]:
%python

df_temp = df_train.join(df_meal_info, on=['meal_id'], how='inner')
df_join = df_temp.join(df_center, on=['center_id'], how='inner')
df_join.columns

Out[8]: ['center_id',
 'meal_id',
 'id',
 'week',
 'checkout_price',
 'base_price',
 'emailer_for_promotion',
 'homepage_featured',
 'num_orders',
 'category',
 'cuisine',
 'city_code',
 'region_code',
 'center_type',
 'op_area']

In [0]:
%python
# Create a view or table

temp_table_name = "jointable"

df_join.createOrReplaceTempView(temp_table_name)

In [0]:
%python
df_join.show()

+---------+-------+-------+----+--------------+----------+---------------------+-----------------+----------+---------+-----------+---------+-----------+-----------+-------+
center_id|meal_id| id|week|checkout_price|base_price|emailer_for_promotion|homepage_featured|num_orders| category| cuisine|city_code|region_code|center_type|op_area|
+---------+-------+-------+----+--------------+----------+---------------------+-----------------+----------+---------+-----------+---------+-----------+-----------+-------+
 55| 1885|1379560| 1| 136.83| 152.29| 0| 0| 177|Beverages| Thai| 647| 56| TYPE_C| 2.0|
 55| 1993|1466964| 1| 136.83| 135.83| 0| 0| 270|Beverages| Thai| 647| 56| TYPE_C| 2.0|
 55| 2539|1346989| 1| 134.86| 135.86| 0| 0| 189|Beverages| Thai| 647| 56| TYPE_C| 2.0|
 55| 2139|1338232| 1| 339.5| 437.53| 0| 0| 54|Beverages| Indian| 647| 56| TYPE_C| 2.0|
 55| 2631|1448490| 1| 243.5| 242.5| 0| 0| 40|Beverages| Indian| 647| 56| TYPE_C| 2.0|
 55| 1248|1270037| 1| 251.23| 252.23| 0| 0| 28|Beverages| Indian| 647| 56| TYPE_C| 2.0|
 55| 1778|1191377| 1| 183.36| 184.36| 0| 0| 190|Beverages| Italian| 647| 56| TYPE_C| 2.0|
 55| 1062|1499955| 1| 182.36| 183.36| 0| 0| 391|Beverages| Italian| 647| 56| TYPE_C| 2.0|
 55| 2707|1025244| 1| 193.06| 192.06| 0| 0| 472|Beverages| Italian| 647| 56| TYPE_C| 2.0|
 55| 1207|1054194| 1| 325.92| 384.18| 0| 1| 676|Beverages|Continental| 647| 56| TYPE_C| 2.0|
 55| 1230|1469367| 1| 323.01| 390.0| 0| 1| 823|Beverages|Continental| 647| 56| TYPE_C| 2.0|
 55| 2322|1029333| 1| 322.07| 388.0| 0| 1| 972|Beverages|Continental| 647| 56| TYPE_C| 2.0|
 55| 2290|1446016| 1| 311.43| 310.43| 0| 0| 162|Rice Bowl| Indian| 647| 56| TYPE_C| 2.0|
 55| 1727|1244647| 1| 445.23| 446.23| 0| 0| 420|Rice Bowl| Indian| 647| 56| TYPE_C| 2.0|
 55| 1109|1378227| 1| 264.84| 297.79| 1| 0| 756|Rice Bowl| Indian| 647| 56| TYPE_C| 2.0|
 55| 2640|1181556| 1| 282.33| 281.33| 0| 0| 108| Starters| Thai| 647| 56| TYPE_C| 2.0|
 55| 2306|1313873| 1| 243.5| 340.53| 0| 0| 28| Pasta| Italian| 647| 56| TYPE_C| 2.0|
 55| 2126|1067069| 1| 486.0| 485.0| 0| 0| 28| Pasta| Italian| 647| 56| TYPE_C| 2.0|
 55| 2826|1058482| 1| 306.58| 305.58| 0| 0| 188| Sandwich| Italian| 647| 56| TYPE_C| 2.0|
 55| 1754|1240935| 1| 289.12| 289.12| 0| 0| 485| Sandwich| Italian| 647| 56| TYPE_C| 2.0|
+---------+-------+-------+----+--------------+----------+---------------------+-----------------+----------+---------+-----------+---------+-----------+-----------+-------+
only showing top 20 rows

-----------

Assignment Questions

Q1. What are the distinct number of meal categories and cuisines?

In [0]:
%python

spark.sql("""

select COUNT(DISTINCT category) as Count_distinct_meal_categories, COUNT(DISTINCT cuisine) as Count_distinct_cuisines
from jointable

""").show()


+------------------------------+-----------------------+
Count_distinct_meal_categories|Count_distinct_cuisines|
+------------------------------+-----------------------+
 14| 4|
+------------------------------+-----------------------+

distinct_meal_categories = 14

distinct_cuisines = 4

----

Q2. Which center_id has the highest num_orders?

In [0]:

%python

spark.sql("""

SELECT center_id, SUM (num_orders) as Sum_order
FROM `jointable`
GROUP BY center_id
order by Sum_order desc 
limit 1

""").show()

+---------+---------+
center_id|Sum_order|
+---------+---------+
 13| 1742220|
+---------+---------+

center_id 13 has the highest num_orders.

----

Q3. What is the top selling cuisine at the center_id that had the highest num_orders?

In [0]:

%python

spark.sql('''

SELECT cuisine, sum (num_orders) as Total_orders
FROM jointable 
WHERE center_id = (
                    SELECT center_id
                    FROM `jointable`
                    GROUP BY center_id
                    order by SUM (num_orders) desc 
                    limit 1
                   )
group by cuisine
Order by Total_orders desc
limit 1

''').show()





+-------+------------+
cuisine|Total_orders|
+-------+------------+
 Thai| 654724|
+-------+------------+

top selling cuisine at the center_id that had the highest num_orders = Thai

----

Q4. What is the average op_area per center_type?

In [0]:
%python

spark.sql('''

SELECT center_type, avg(op_area)
FROM jointable 
group by center_type

''').show()



+-----------+------------------+
center_type| avg(op_area)|
+-----------+------------------+
 TYPE_C|3.2724664576805202|
 TYPE_B| 4.881073454067083|
 TYPE_A| 4.109563092841962|
+-----------+------------------+

average op_area per center_type 

Type_C = 3.2724664576805202

TYPE_B = 4.881073454067083

TYPE_A = 4.109563092841962

----

Q5. Which center_type had the highest revenue? (Revenue = total sum of checkout_price*num_orders)

In [0]:
%python

spark.sql('''

SELECT center_type, sum(checkout_price*num_orders) as Revenue
FROM jointable 
group by center_type
Order by Revenue desc
limit 1

''').show()



+-----------+-------------------+
center_type| Revenue|
+-----------+-------------------+
 TYPE_A|7.276203201870064E9|
+-----------+-------------------+


center_type that had the highest revenue =  Type_A

----

Q6. Which is the top ordered cuisine in terms of num_orders?

In [0]:
%python

spark.sql('''

SELECT cuisine, sum(num_orders) as Total_order
FROM jointable 
group by cuisine
Order by Total_order desc
limit 1

''').show()



+-------+-----------+
cuisine|Total_order|
+-------+-----------+
Italian| 17166334|
+-------+-----------+

top ordered cuisine in terms of num_orders = Italian

----

Q7. What are the num_orders per cuisine per week?

In [0]:

%python

spark.sql('''

SELECT cuisine, week , sum (num_orders) as total_orders
FROM jointable 
group by cuisine, week
Order by cuisine, week

''').show()



+-----------+----+------------+
 cuisine|week|total_orders|
+-----------+----+------------+
Continental| 1| 146020|
Continental| 2| 133570|
Continental| 3| 97977|
Continental| 4| 118819|
Continental| 5| 116077|
Continental| 6| 107160|
Continental| 7| 77051|
Continental| 8| 43897|
Continental| 9| 152678|
Continental| 10| 156051|
Continental| 11| 124520|
Continental| 12| 95914|
Continental| 13| 109010|
Continental| 14| 98188|
Continental| 15| 96401|
Continental| 16| 90087|
Continental| 17| 114494|
Continental| 18| 107330|
Continental| 19| 80208|
Continental| 20| 87718|
+-----------+----+------------+
only showing top 20 rows

----

Q8. Which center_id gave the highest number of discounts? (Discount is considered when checkout_price is less than base_price)

In [0]:

%python

spark.sql("""

SELECT  center_id, Count (CASE
                                                                   WHEN (jointable.checkout_price < jointable.base_price)
                                                                   THEN 1
                                                                   ElSE NULL
                                                                   END ) AS Count_of_Discounts

FROM jointable
group by center_id
Order by Count_of_Discounts desc
limit 1

""").show()



+---------+------------------+
center_id|Count_of_Discounts|
+---------+------------------+
 13| 1509|
+---------+------------------+

center_id 13 has the highest number of discounts.